core/src/config/mod.rs

```diff
@@ -97,15 +98,18 @@ impl<T> BlockHash for T where
 /// and extrinsics.
 pub trait Hasher {
     /// The type given back from the hash operation
-    type Output;
+    type Output: Hash;
+
+    /// Construct a new hasher.
+    fn new(metadata: &Metadata) -> Self;
 
     /// Hash some bytes to the given output type.
-    fn hash(s: &[u8]) -> Self::Output;
+    fn hash(&self, s: &[u8]) -> Self::Output;
 
```

Comment:
@niklasad1 this is the important change; we can now construct a hasher given metadata (which might be expensive), and then `hash` takes `&self` now so we can hash based on this metadata. This is why we pass a hasher around more. We _could_ construct it any place we have access to metadata, but the point is to avoid this because it could be expensive